File: problems/leetcode/84.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        stack<int> stck;
        int max_area = 0, n = heights.size();

        for (int i = 0; i < n; i++) {
            while(!stck.empty() && heights[i] < heights[stck.top()]) {
                int temp_height = stck.top();
                stck.pop();
                int temp = stck.empty() ? -1 : stck.top();
                int start_i = i;
                max_area = max(max_area, heights[temp_height] * (start_i - temp - 1));
            }
            stck.push(i);
        }
        while (!stck.empty()) {
            int temp_height = stck.top();
                stck.pop();
                int temp = stck.empty() ? -1 : stck.top();
                int start_i = n;
                max_area = max(max_area, heights[temp_height] * (start_i - temp - 1));
        }

        return max_area;

    }
};

File: problems/leetcode/84.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stck = []
        max_area = 0

        for i, h in enumerate(heights):
            start_i = i
            while stck and stck[-1][1] > h:
                temp_index, temp_height = stck.pop()
                max_area = max(max_area, temp_height * (i - temp_index))
                start_i = temp_index
            stck.append((start_i, h))
        
        for i, h in stck:
            max_area = max(max_area, h * (len(heights) - i))
        return max_area